document.go

```diff
@@ -0,0 +1,52 @@
+package smithy
+
+// A document type represents an untyped JSON-like value.
+//
+// Not all protocols support document types, and the serialization format of a
+// document type is protocol specific. All JSON protocols SHOULD support
+// document types and they SHOULD serialize document types inline as normal
+// JSON values.
+type Document interface {
+	// Marshal will attempt to serialize the provided type into a
+	// protocol specific document value.
+	Marshal(x interface{}) error
+
+	// Unmarshall will attempt to map the protocol specific document value
+	// onto the provided type. If the type does not match the document type
+	// an error will be returned.
+	Unmarshal(x interface{}) error
```

Comment:
To prevent potentially colliding with other `Unmarshal` methods of other types it would be helpful to rename these to `MarshalDocument`, `UnmarshalDocument`. This would also follow the common pattern in the stdlib for serialization methods.